components/raftstore/src/store/compaction_guard.rs

```diff
@@ -153,27 +182,52 @@ impl<P: RegionInfoProvider> SstPartitioner for CompactionGuardGenerator<P> {
         if !self.use_guard {
             return SstPartitionerResult::NotRequired;
         }
-        let mut pos = self.pos;
-        let mut skip_count = 0;
-        while pos < self.boundaries.len() && self.boundaries[pos].as_slice() <= req.prev_user_key {
-            pos += 1;
-            skip_count += 1;
-            if skip_count >= COMPACTION_GUARD_MAX_POS_SKIP {
-                let prev_user_key = req.prev_user_key.to_vec();
-                pos = match self.boundaries.binary_search(&prev_user_key) {
-                    Ok(search_pos) => search_pos + 1,
-                    Err(search_pos) => search_pos,
-                };
-                break;
-            }
+        self.pos = seek_to(&self.boundaries, req.prev_user_key, self.pos);
+        // Generally this shall be a noop... because each time we are moving the cursor
+        // to the previous key.
+        let left_next_level_pos = seek_to(
+            &self.next_level_boundaries,
+            req.prev_user_key,
+            self.next_level_pos,
+        );
+        let right_next_level_pos = seek_to(
+            &self.next_level_boundaries,
+            req.current_user_key,
+            left_next_level_pos,
+        );
+        // The cursor has been moved.
+        if right_next_level_pos > left_next_level_pos {
+            self.current_next_level_size += self.next_level_size
+                [left_next_level_pos..right_next_level_pos - 1]
```

Comment:
Why left_next_level_pos does **not** minus 1, but right_next_level_pos does.

Reply:
```
   0   1   2   3   4 <- next_level_boundaries
___|___|***|***|___|
     0   1   2   3   <- next_level_size
     ^           ^
```

Somehow this is a negative strategy -- we ignore all segments the key partially crossed. Given the segment size << the max compaction size, this may not make things worse.

The diagram above shows, currently we checked the size of the stared segments. Perhaps involving the segment `0` (i.e. minus `1` to `left_next_level_pos`) can help us get a more accurate statistic.